ldap_auth_provider.py

```diff
@@ -84,6 +84,7 @@ def __init__(self, config, account_handler):
         self.ldap_active_directory = config.active_directory
         if self.ldap_active_directory:
             self.ldap_default_domain = config.default_domain
+            defer.ensureDeferred(self._fetch_root_domain())
```

Comment:
this won't wait for the `_fetch_root_domain` operation to complete, so this will give erratic behaviour.

In general, you can't really use asynchronous operations from within a constructor. I suggest the easiest solution is going to be to perform this logic in the first call to `check_auth`. However, since it's only applicable for SEARCH mode, it might be best to first refactor `check_auth` to separate the SEARCH and SIMPLE cases into separate methods.

Reply:
Maybe make ```_fetch_root_domain``` with all other calls sync?
We really need async init here.

Reply:
making synapse's startup dependent on the ldap server being reachable sounds unsatisfactory, to be honest.

why is it so important to do this during init?

Reply:
> making synapse's startup dependent on the ldap server being reachable sounds unsatisfactory, to be honest.

You are right, nvm